setwd("~/Documents/nathansonlab/CNV/final")
# pluta 9/9/21


# ---------------------------------------------------------------- #
exprCNVBoxPlot <- function( cnv, expr, cpg, GENENAME)
# create the expr by CNV box plot
{
  
  
  dat <- data.frame(expr = expr, 
                    cnv = cnv, 
                    cpg = cpg)
  
  if(any(dat$expr == 0))
  {
    dat  <- dat[ dat$expr != 0, ]
  }
  
  # plot CNV against expression after correcting for the effect of cpg methylation
  fit1 <- lm( data = dat, log2(expr) ~ cpg)
  fit2 <- lm( data = dat, log2(expr) ~ as.integer(cnv) + cpg)
  
  dat$cnv <- as.factor(dat$cnv)
 
  
  p1 <- ggplot(data = dat, aes(x = cnv, y  = fit1$residuals, fill = cnv)) +
    geom_boxplot() + 
    theme_minimal() +
    labs(y = "log2(FPKM)",
         title = GENENAME,
         subtitle = paste0("p = ", round(summary(fit2)$coefficients[2,4], 3)))
  
  return(p1)
}
# ---------------------------------------------------------------- #


# ---------------------------------------------------------------- #
formatData <-  function( dat, type )
{
  dat  <- t(dat)
  n <- dim(dat)[1]
  m <- dim(dat)[2]
  
  colnames(dat)  <- c(dat[n,1:m])
  dat <- dat[-which(rownames(dat) %in%  "GeneName"),]
  
  if( type ==  "cnv" )
  {
    dat <- apply(dat, 2, as.integer)
  } else
  {
    dat <- apply(dat, 2, as.numeric)
  }
  
  return(dat)
}
# ---------------------------------------------------------------- #





# ================================================================ #
# ============================ MAIN ============================== #
# ================================================================ #
# matrices generated by getTCGAMatrices.R
# contain gene-level CNV, gene expression, and cpg methylation, all from TCGA
cnv  <- formatData(read.table("TCGAcnvMatrix.txt", header =  TRUE), "cnv")
expr <- formatData(read.table("TCGAexpressionMatrix.txt", header  = TRUE), "expr")
cpg  <- formatData(read.table("TCGAcpgmethMatrix.txt", header = TRUE), "cpg")

# need to check that all matrices are in the same order
expr <- expr[ ,match(colnames(cnv), colnames(expr))]
cpg <- cpg[ ,match(colnames(cnv), colnames(cpg))]

if( !( all(colnames(cnv) == colnames(expr)) ))
{
  if( !(all(colnames(cnv) == colnames(cpg))))
  {
    stop("colname mismatch")
  }
}
       
# create plots
for( i in 1:dim(cnv)[2])
{
  genename <- colnames(cnv)[i]
  
  p1 <- exprCNVBoxPlot( cnv[,i], expr[,i], cpg[,i], genename)
  
  png(paste0(genename, "_cnv_expr.png"))
  print(p1)
  dev.off()
}

